File: packages/flutter/lib/src/verification_controller.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io' show HttpException;

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

import 'verification.dart';
import 'verification_status.dart';

/// Snapshot of the verification + countdown shown to the user on
/// each tick. Held by [VerificationController] (a [ValueNotifier]);
/// the widget binds to it via [ValueListenableBuilder].
@immutable
class VerificationState {
  final Verification verification;
  final int secondsLeft;
  const VerificationState({required this.verification, required this.secondsLeft});

  @override
  bool operator ==(Object other) =>
      other is VerificationState &&
      other.verification == verification &&
      other.secondsLeft == secondsLeft;

  @override
  int get hashCode => Object.hash(verification, secondsLeft);
}

/// State machine driving the SYROTP verification lifecycle on Flutter.
/// Polls `${baseUrl}/v/:id/status` (the public, IP-rate-limited
/// endpoint shipped in SYROTP server v0.5.0), runs a 1Hz countdown,
/// and emits a fresh [VerificationState] on every visible change.
///
/// Lifecycle: call [start] once; the controller manages its own
/// timers and HTTP calls. Call [dispose] on teardown to cancel
/// timers and close the HTTP client.
///
/// The constructor takes a `now` clock function for testability —
/// pass a fake clock to drive the local TTL fallback path
/// deterministically.
class VerificationController extends ValueNotifier<VerificationState> {
  final String baseUrl;
  final Duration pollInterval;
  final http.Client _http;
  final void Function(Verification) onVerified;
  final void Function(Verification) onExpired;
  final void Function(Verification) onCancelled;
  final void Function(Object) onError;
  final DateTime Function() _now;

  Timer? _pollTimer;
  Timer? _countdownTimer;
  bool _started = false;
  bool _disposed = false;
  VerificationStatus _prevStatus;

  VerificationController({
    required this.baseUrl,
    required Verification initial,
    this.pollInterval = const Duration(milliseconds: 2500),
    http.Client? httpClient,
    void Function(Verification)? onVerified,
    void Function(Verification)? onExpired,
    void Function(Verification)? onCancelled,
    void Function(Object)? onError,
    DateTime Function()? now,
  })  : _http = httpClient ?? http.Client(),
        onVerified = onVerified ?? _noopV,
        onExpired = onExpired ?? _noopV,
        onCancelled = onCancelled ?? _noopV,
        onError = onError ?? _noopE,
        _now = now ?? DateTime.now,
        _prevStatus = initial.status,
        super(VerificationState(
          verification: initial,
          secondsLeft: _calcSecondsLeft(initial.expiresAt, now ?? DateTime.now),
        ));

  /// Start polling + countdown. Idempotent — calling start() a
  /// second time is a no-op.
  void start() {
    if (_started || _disposed) return;
    _started = true;
    if (value.verification.status.isTerminal) return;
    _countdownTimer = Timer.periodic(const Duration(seconds: 1), (_) => _onTick());
    // Fire one immediate poll so a fast-completing verification
    // surfaces without waiting a full interval.
    unawaited(_pollOnce());
    _pollTimer = Timer.periodic(pollInterval, (_) => unawaited(_pollOnce()));
  }

  @override
  void dispose() {
    _disposed = true;
    _pollTimer?.cancel();
    _countdownTimer?.cancel();
    _http.close();
    super.dispose();
  }

  void _onTick() {
    if (_disposed) return;
    final v = value.verification;
    if (v.status.isTerminal) return;
    final secs = _calcSecondsLeft(v.expiresAt, _now);
    if (secs <= 0) {
      _transition(v.withTransition(
        status: VerificationStatus.expired,
        expiresAt: v.expiresAt,
      ));
      return;
    }
    if (secs != value.secondsLeft) {
      value = VerificationState(verification: v, secondsLeft: secs);
    }
  }

  Future<void> _pollOnce() async {
    if (_disposed) return;
    final v = value.verification;
    if (v.status.isTerminal) return;
    final url =
        '${baseUrl.replaceAll(RegExp(r'/+$'), '')}/v/${Uri.encodeComponent(v.id)}/status';
    try {
      final res = await _http.get(
        Uri.parse(url),
        headers: const {'Accept': 'application/json'},
      );
      if (_disposed) return;
      if (res.statusCode < 200 || res.statusCode >= 300) {
        throw HttpException('status poll failed: HTTP ${res.statusCode}');
      }
      final body = jsonDecode(res.body) as Map<String, dynamic>;
      final newStatus = VerificationStatus.fromWire(body['status'] as String);
      final newExpiresAt = DateTime.parse(body['expires_at'] as String);
      final verifiedRaw = body['verified_at'];
      final newVerifiedAt =
          verifiedRaw == null ? null : DateTime.parse(verifiedRaw as String);

      final current = value.verification;
      if (newStatus == current.status) {
        if (newExpiresAt != current.expiresAt) {
          final updated = current.withExpiresAt(newExpiresAt);
          value = VerificationState(
            verification: updated,
            secondsLeft: _calcSecondsLeft(newExpiresAt, _now),
          );
        }
        return;
      }
      _transition(current.withTransition(
        status: newStatus,
        expiresAt: newExpiresAt,
        verifiedAt: newVerifiedAt,
      ));
    } catch (e) {
      if (_disposed) return;
      onError(e);
    }
  }

  void _transition(Verification next) {
    final prev = _prevStatus;
    _prevStatus = next.status;
    value = VerificationState(
      verification: next,
      secondsLeft: _calcSecondsLeft(next.expiresAt, _now),
    );
    if (prev != next.status && prev == VerificationStatus.pending) {
      switch (next.status) {
        case VerificationStatus.verified:
          onVerified(next);
          break;
        case VerificationStatus.expired:
          onExpired(next);
          break;
        case VerificationStatus.cancelled:
          onCancelled(next);
          break;
        case VerificationStatus.pending:
        case VerificationStatus.failed:
          break;
      }
    }
    if (next.status.isTerminal) {
      _pollTimer?.cancel();
      _countdownTimer?.cancel();
      _pollTimer = null;
      _countdownTimer = null;
    }
  }

  static int _calcSecondsLeft(DateTime expiresAt, DateTime Function() now) {
    final diff = expiresAt.difference(now()).inSeconds;
    return diff < 0 ? 0 : diff;
  }

  static void _noopV(Verification _) {}
  static void _noopE(Object _) {}
}

File: packages/android-ui/library/src/main/kotlin/dev/orotp/ui/VerificationController.kt
package dev.syrotp.ui

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.time.Instant
import java.util.concurrent.TimeUnit

/**
 * Snapshot of the verification + countdown shown to the user on
 * each tick. Held by the controller's [StateFlow]; the Compose
 * layer collects this as state.
 */
data class VerificationState(
    val verification: Verification,
    val secondsLeft: Long,
)

/**
 * Framework-agnostic state machine driving the SYROTP verification
 * lifecycle on Android. Polls `${baseUrl}/v/:id/status` (the
 * public, IP-rate-limited endpoint shipped in SYROTP server v0.5.0),
 * runs a 1Hz countdown, and emits a fresh [VerificationState] on
 * every visible change.
 *
 * Designed to live independent of Compose so unit tests can run on
 * the JVM without Robolectric or instrumented tests. The contract
 * is intentionally identical to `useSyrotpVerification` in
 * `@syrotp/react` and `VerificationController` in
 * `@syrotp/web-component`: same URL shape, same merge rules, same
 * local TTL fallback.
 *
 * Lifecycle:
 *   - [start] launches polling + countdown coroutines as children
 *     of the supplied parent scope. Calling start a second time is
 *     a no-op.
 *   - [stop] cancels the scope. Safe to call from
 *     `DisposableEffect.onDispose`.
 *
 * The constructor takes a `nowMs` clock for testability — pass a
 * fake clock to drive the local TTL fallback path deterministically
 * in unit tests.
 */
class VerificationController(
    private val baseUrl: String,
    initial: Verification,
    private val pollIntervalMs: Long = DEFAULT_POLL_INTERVAL_MS,
    private val httpClient: OkHttpClient = defaultHttpClient(),
    private val onVerified: (Verification) -> Unit = {},
    private val onExpired: (Verification) -> Unit = {},
    private val onCancelled: (Verification) -> Unit = {},
    private val onError: (Throwable) -> Unit = {},
    private val nowMs: () -> Long = { System.currentTimeMillis() },
) {
    private val _state = MutableStateFlow(
        VerificationState(
            verification = initial,
            secondsLeft = secondsLeftFrom(initial.expiresAt, nowMs()),
        ),
    )
    val state: StateFlow<VerificationState> = _state.asStateFlow()

    private var scope: CoroutineScope? = null

    fun start(parent: CoroutineScope) {
        if (scope != null) return
        val s = CoroutineScope(parent.coroutineContext + SupervisorJob())
        scope = s
        if (_state.value.verification.status.isTerminal) return
        s.launch { runCountdown() }
        s.launch { runPolling() }
    }

    fun stop() {
        scope?.cancel()
        scope = null
    }

    private suspend fun runCountdown() {
        while (scopeActive()) {
            val v = _state.value.verification
            if (v.status.isTerminal) return
            val secs = secondsLeftFrom(v.expiresAt, nowMs())
            if (secs <= 0L) {
                transition(
                    v.copy(
                        status = VerificationStatus.Expired,
                        sendTo = null,
                        message = null,
                    ),
                )
                return
            }
            if (secs != _state.value.secondsLeft) {
                _state.value = _state.value.copy(secondsLeft = secs)
            }
            delay(COUNTDOWN_TICK_MS)
        }
    }

    private suspend fun runPolling() {
        // Fire one immediate poll so a fast-completing verification
        // surfaces without waiting a full interval.
        pollOnce()
        while (scopeActive()) {
            delay(pollIntervalMs)
            if (_state.value.verification.status.isTerminal) return
            pollOnce()
        }
    }

    private suspend fun pollOnce() {
        val v = _state.value.verification
        if (v.status.isTerminal) return
        val url = "${baseUrl.trimEnd('/')}/v/${v.id}/status"
        val req = Request.Builder()
            .url(url)
            .header("Accept", "application/json")
            .get()
            .build()
        try {
            val raw = withContext(Dispatchers.IO) {
                httpClient.newCall(req).execute().use { res ->
                    if (!res.isSuccessful) {
                        throw IOException("status poll failed: HTTP ${res.code}")
                    }
                    res.body?.string() ?: throw IOException("empty body")
                }
            }
            val parsed = jsonCodec.decodeFromString<StatusResponse>(raw)
            val current = _state.value.verification
            if (parsed.status == current.status) {
                if (parsed.expiresAt != current.expiresAt) {
                    val updated = current.copy(expiresAt = parsed.expiresAt)
                    _state.value = VerificationState(
                        verification = updated,
                        secondsLeft = secondsLeftFrom(updated.expiresAt, nowMs()),
                    )
                }
                return
            }
            transition(
                current.copy(
                    status = parsed.status,
                    expiresAt = parsed.expiresAt,
                    verifiedAt = parsed.verifiedAt,
                    sendTo = if (parsed.status == VerificationStatus.Pending) current.sendTo else null,
                    message = if (parsed.status == VerificationStatus.Pending) current.message else null,
                ),
            )
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            onError(e)
        }
    }

    private fun transition(next: Verification) {
        val prev = _state.value.verification.status
        _state.value = VerificationState(
            verification = next,
            secondsLeft = secondsLeftFrom(next.expiresAt, nowMs()),
        )
        if (prev != next.status && prev == VerificationStatus.Pending) {
            when (next.status) {
                VerificationStatus.Verified -> onVerified(next)
                VerificationStatus.Expired -> onExpired(next)
                VerificationStatus.Cancelled -> onCancelled(next)
                else -> Unit
            }
        }
    }

    private fun scopeActive(): Boolean = scope?.isActive == true

    @Serializable
    private data class StatusResponse(
        val status: VerificationStatus,
        @SerialName("expires_at") val expiresAt: String,
        @SerialName("verified_at") val verifiedAt: String? = null,
    )

    companion object {
        const val DEFAULT_POLL_INTERVAL_MS: Long = 2500L
        private const val COUNTDOWN_TICK_MS: Long = 1000L

        private val jsonCodec = Json { ignoreUnknownKeys = true }

        private fun defaultHttpClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .readTimeout(5, TimeUnit.SECONDS)
            .build()

        internal fun secondsLeftFrom(expiresAt: String, nowMs: Long): Long {
            val expiresMs = try {
                Instant.parse(expiresAt).toEpochMilli()
            } catch (_: Throwable) {
                return 0L
            }
            return ((expiresMs - nowMs) / 1000L).coerceAtLeast(0L)
        }
    }
}
